lib/orderbook/OrderBook.ts

```diff
@@ -1,130 +1,205 @@
-      this.repository.getOwnOrders(),
-    ]);
-
-    const peerBuyOrdersByPairs = groupBy(peerOrders.buyOrders, order => order.pairId);
-    const peerSellOrdersByPairs = groupBy(peerOrders.sellOrders, order => order.pairId);
-    const ownBuyOrdersByPairs = groupBy(ownOrders.buyOrders, order => order.pairId);
-    const ownSellOrdersByPairs = groupBy(ownOrders.sellOrders, order => order.pairId);
+  public init = async () => {
+    const pairs = await this.repository.getPairs();
 
     pairs.forEach((pair) => {
       this.matchingEngines[pair.id] = new MatchingEngine(
         pair.id,
         this.config.internalmatching,
-        peerBuyOrdersByPairs[pair.id],
-        peerSellOrdersByPairs[pair.id],
-        ownBuyOrdersByPairs[pair.id],
-        ownSellOrdersByPairs[pair.id],
       );
+      this.ownOrders[pair.id] = this.initOrders();
+      this.peerOrders[pair.id] = this.initOrders();
     });
 
     this.pairs = pairs;
   }
 
+  private initOrders = (): Orders => {
+    return {
+      buyOrders: [],
+      sellOrders: [],
+    };
+  }
+
   /**
    * Returns the list of available trading pairs.
    */
-  public getPairs(): Promise<db.PairInstance[]> {
+  public getPairs = (): Promise<db.PairInstance[]> => {
     return this.repository.getPairs();
   }
 
   /**
    * Returns lists of buy and sell orders of peers sorted by price.
    */
-  public getPeerOrders(pairId: string, maxResults: number): Promise<Orders> {
-    return this.repository.getPeerOrders(pairId, maxResults);
+  public getPeerOrders = (pairId: string, maxResults: number): Orders => {
+    return this.getOrders(maxResults, this.peerOrders[pairId]);
   }
 
   /*
   * Returns lists of the node's own buy and sell orders sorted by price.
   */
-  public getOwnOrders(pairId: string, maxResults: number): Promise<Orders> {
-    return this.repository.getOwnOrders(pairId, maxResults);
+  public getOwnOrders = (pairId: string, maxResults: number): Orders => {
+    return this.getOrders(maxResults, this.ownOrders[pairId]);
+  }
+
+  private getOrders = (maxResults: number, orders: Orders) => {
+    if (maxResults > 0) {
+      const result = this.initOrders();
+
+      result.buyOrders.push(...orders.buyOrders.slice(0, maxResults));
+      result.sellOrders.push(...orders.sellOrders.slice(0, maxResults));
+
+      return result;
+    } else {
+      return orders;
+    }
   }
 
-  public addLimitOrder = async (order: orders.OwnOrder): Promise<matchingEngine.MatchingResult>  => {
+  public addLimitOrder = (order: orders.OwnOrder): matchingEngine.MatchingResult => {
     return this.addOwnOrder(order);
   }
 
-  public addMarketOrder = async (_order: orders.OwnOrder) => {
+  public addMarketOrder = (_order: orders.OwnOrder) => {
     // TODO: implement
   }
 
-  private addOwnOrder = async (order: orders.OwnOrder): Promise<matchingEngine.MatchingResult>  => {
+  public castToOutgoingOrder = async (ownOrders: StampedOrder[], outgoingOrders: OutgoingOrder[]) => {
```

Comment:
This is too similar to `broadcastOrder`, although was implemented differently. I'd rather have the logic of adding an invoice and casting to an `OutgoingOrder` implemented only once.